ethpm/tools/builder.py

```diff
@@ -0,0 +1,332 @@
+        )
+
+    if "sources" not in manifest:
+        manifest["sources"] = {}
+
+    return assoc(
+        manifest,
+        "sources",
+        assoc(manifest["sources"], "./{0}".format(source_path_suffix), source_data),
+    )
+
+
+def pin_source(
+    name: str,
+    compiler_output: Dict[str, Any],
+    ipfs_backend: BaseIPFSBackend,
+    package_root_dir: str = None,
+) -> Manifest:
+    """
+    Pins source to IPFS and returns a copy of manifest with added field to
+    "sources": {relative_source_path: IFPS URI}.
+
+    If `package_root_dir` is not provided, cwd is expected to resolve the relative path
+    to the source as defined in the compiler output.
+    """
+    return _pin_source(name, compiler_output, ipfs_backend, package_root_dir)
+
+
+@curry
+def _pin_source(
+    name: str,
+    compiler_output: Dict[str, Any],
+    ipfs_backend: BaseIPFSBackend,
+    package_root_dir: str,
+    manifest: Manifest,
+) -> Manifest:
+    names_and_paths = get_names_and_paths(compiler_output)
+    source_path = names_and_paths[name]
+    if package_root_dir:
+        (ipfs_data,) = ipfs_backend.pin_assets(package_root_dir / source_path)
+    else:
+        cwd = Path(os.getcwd())
+        (ipfs_data,) = ipfs_backend.pin_assets(cwd / source_path)
+
+    if "sources" not in manifest:
+        manifest["sources"] = {}
+
+    return assoc(
+        manifest,
+        "sources",
+        assoc(
+            manifest["sources"],
+            "./{0}".format(source_path),
+            "ipfs://{0}".format(ipfs_data["Hash"]),
+        ),
+    )
+
+
+#
+# Contract Types
+#
+
+
+def contract_type(name: str, compiler_output: Dict[str, Any], **kwargs) -> Manifest:
+    """
+    Returns a copy of manifest with added contract_data field as specified by kwargs.
+    If no kwargs are present, all available contract_data found in the compiler output
+    will be included.
+
+    To include specific contract_data fields, add kwarg set to True (i.e. `abi=True`)
+    To alias a contract_type, include a kwarg `alias` (i.e. `alias="OwnedAlias"`)
+    If only an alias kwarg is provided, all available contract data will be included.
+    Kwargs must match fields as defined in the EthPM Spec (except "alias") if user
+        wants to include them in custom contract_type.
+    """
+    return _contract_type(name, compiler_output, kwargs)
+
+
+@curry
+def _contract_type(
+    name: str, compiler_output: Dict[str, Any], kwargs, manifest: Manifest
+) -> Manifest:
+    contracts_by_name = normalize_compiler_output(compiler_output)
+    try:
+        all_type_data = contracts_by_name[name]
+    except KeyError:
+        raise ManifestBuildingError(
+            "Contract name: {0} not found in the provided compiler output.".format(name)
+        )
+
+    if not kwargs:
+        return assoc_in(manifest, ["contract_types", name], all_type_data)
+
+    # Handle single kwarg if "alias"
+    if "alias" in kwargs and len(kwargs) is 1:
```

Comment:
Maybe remove this from `kwargs` and make `alias` an independent argument that defaults to `None`?